### PR TITLE
dlt: Adaptation for QNX build system with SDP 7.x

### PR DESCRIPTION
**Pull request by marcin-sochacki**
Hello Team,
I would like to contribute to the dlt-daemon project with the following changes:
- support for more flexible installation (in sysroot or in a staging area)
- support for installation of the units tests and relevant files (this helps to execute cross-compiled unit tests)
- support for recursive QNX makefiles (added qnx folder)
The changes are split in 3 commits for clarity (you need all of them).

Building dlt-daemon project for QNX:
```
source sdp_directory/qnxsdp-env.sh
make -C qnx/build JLEVEL=4
```
This would compile the code for aarch64le and x86_64 architectures (output files are located in qnx/build folder).
The build system will utilize (in the above example) 4 processor cores.

You can also build and install the project:

`make -C qnx/build install JLEVEL=4`
It will install the binaries in sysroot (SDP folder)

`make -C qnx/build install JLEVEL=4 INSTALL_ROOT_nto=PATH_TO_YOUR_STAGING_AREA USE_INSTALL_ROOT=true`
It will install the binaries in a staging area (a folder of your choice)

Prefixing the command with CPULIST=x86_64 or with CPULIST=aarch64 wil build only specified architecture, for example:
`CPULIST=x86_64 make -C qnx/build JLEVEL=4`

Best Regards
Marcin Sochacki

P.S.
You need a QNX license for compiling the code.
New "QNX Everywhere" initiative will be introduced soon and provide licenses for a non commercial use of the QNX ecosystem:
https://www.blackberry.com/us/en/company/newsroom/press-releases/2024/blackberry-launches-qnx-everywhere-to-address-global-embedded-software-developer-skills-shortage

### CODE REVIEW

**Comment by minminlittleshrimp** [comment]
Hello @marcin-sochacki 
nice contribution!
I am also excited hearing "QNX Everywhere" will be introduced soon, that is a great new for Embedded SW guys to dig deep into Microkernel QNX Neutrino. However, due to the big scale of this PR, I am not so sure if @michael-methner can suggest any good idea to process this CRQ. Please kindly check Michael, thank you.
Anyway, looking forward to developing DLT under Microkernel Free License.
Happy coding!

**Comment by marcin-sochacki** [comment]
@michael-methner 

The first 2 commits are OS agnostic. They re-factor CMake to allow installation to a staging area and to support installation of the unit tests.

The third commit provides QNX recursive makefiles for building QNX binary artifacts for arm and for x86 architectures with a predefined set of the CMake variables. These files are located in a qnx subfolder and this way isolated from the other files.

I would like to support you with extending Github Actions to run QNX builds and tests as soon as it will be possible. For now I think we could live with possible undetected regression (in regard to QNX builds).

**Comment by marcin-sochacki** [comment]
FYI:
QNX port of the version v2.18.10 was made available in the QNX public area on the github:
https://github.com/qnx/dlt-daemon/tree/qnx-v2.18.10

This pull request is still valid, since we would like merge the changes to the master branch to avoid unnecessary efforts with porting upcoming releases.

**Comment by duvanan13** [comment]
Hello

**Comment by minminlittleshrimp** [comment]
> Hello

Hello An,
thanks for joining this topic.
The topic is very interesting for learning and developing.
Let's work together!

**Comment by marcin-sochacki** [comment]
Any chances to merge this PR?
I wanted to start working on https://github.com/COVESA/dlt-daemon/pull/345, but it would help me to get the changes upstreamed first.

**Comment by minminlittleshrimp** [comment]
Hello @duvanan13 
what is the status of this PR?
Did you try:
- Setup QNX build system local?
- Do module build with dlt?
- Check if DLT can be added into QNX image?
- Flash it on a target/RTA node/server?
- Or, run qemu?
Please kindly update your side.

**Comment by marcin-sochacki** [comment]
Hello @duvanan13, Hello @minminlittleshrimp

the commits I asked for upstreaming are a bare minimum (no testing scripts, etc.).
I didn't want to push QNX internal stuff, which may not be interesting to a project.

However for a local testing you may find the following scripts helpful:
https://github.com/qnx/dlt-daemon/tree/v2.18.10_QNX/qnx/test
They were created for v2.18.10 based on tests/CMakeLists.txt.
You can basically copy qnx/test folder to your repository and follow the instructions.

Testing is based on qemu (you have to install qemu first). 
The following test suites are executed:
gtest_dlt_common
gtest_dlt_user
gtest_dlt_daemon_common
dlt_env_ll_unit_test
gtest_dlt_daemon_gateway
gtest_dlt_daemon_offline_log
gtest_dlt_daemon_event_handler
gtest_dlt_daemon_multiple_files_logging

All together 598 tests. Only one of them fails:
t_dlt_user_shutdown_while_init_is_running (from gtest_dlt_user)
I created an issue for it, since it is not a QNX specific problem (it is reproducible under Linux as well).
https://github.com/COVESA/dlt-daemon/issues/522

Cheers

**Comment by minminlittleshrimp** [comment]
Hello @marcin-sochacki 
that document is really helpful.

Please @duvanan13 follow the doc and setup to test in your local.
Soon we need to review this PR and at least now I know that we can test by one way or another to see it works.
Thanks

**Comment by minminlittleshrimp** [comment]
Status:
Facing linking issue:
```
[ 68%] Linking CXX executable dlt-qnx-system
cd /home/lum3hc/work/github/qnx/dlt-daemon/qnx/build/nto-aarch64-le/build/src/dlt-qnx-system && /usr/bin/cmake -E cmake_link_script CMakeFiles/dlt-qnx-system.dir/link.txt --verbose=1
/home/lum3hc/qnx_tools/qnx-sdp/host/linux/x86_64/usr/bin/q++ -Vgcc_ntoaarch64le -lang-c++   -I. -I/home/lum3hc/work/github/qnx/dlt-daemon/qnx/build/nto-aarch64-le -I/home/lum3hc/work/github/qnx/dlt-daemon/qnx/build -I/home/lum3hc/qnx_tools/qnx-sdp/target/qnx7/usr/include   -I/home/lum3hc/qnx_tools/qnx-sdp/target/qnx7/aarch64le//usr/local/include -O3 -DNDEBUG -O3 -DNDEBUG  -Wl,--build-id=md5 -Wl,--allow-shlib-undefined CMakeFiles/dlt-qnx-system.dir/dlt-qnx-system.c.o CMakeFiles/dlt-qnx-system.dir/dlt-qnx-slogger2-adapter.cpp.o  -o dlt-qnx-system  -Wl,-rpath,/home/lum3hc/work/github/qnx/dlt-daemon/qnx/build/nto-aarch64-le/build/src/lib: ../lib/libdlt.so.2 -lslog2parse -ljson -lsocket 
CMakeFiles/dlt-qnx-system.dir/dlt-qnx-system.c.o: In function `main':
dlt-qnx-system.c:(.text.startup+0x264): undefined reference to `strndup'
dlt-qnx-system.c:(.text.startup+0x2e0): undefined reference to `strndup'
dlt-qnx-system.c:(.text.startup+0x67c): undefined reference to `strndup'
cc: /home/lum3hc/qnx_tools/qnx-sdp/host/linux/x86_64/usr/bin/aarch64-unknown-nto-qnx7.0.0-ld error 1
src/dlt-qnx-system/CMakeFiles/dlt-qnx-system.dir/build.make:99: recipe for target 'src/dlt-qnx-system/dlt-qnx-system' failed
make[4]: *** [src/dlt-qnx-system/dlt-qnx-system] Error 1
make[4]: Leaving directory '/home/lum3hc/work/github/qnx/dlt-daemon/qnx/build/nto-aarch64-le/build'
CMakeFiles/Makefile2:1465: recipe for target 'src/dlt-qnx-system/CMakeFiles/dlt-qnx-system.dir/all' failed
make[3]: *** [src/dlt-qnx-system/CMakeFiles/dlt-qnx-system.dir/all] Error 2
make[3]: Leaving directory '/home/lum3hc/work/github/qnx/dlt-daemon/qnx/build/nto-aarch64-le/build'
Makefile:140: recipe for target 'all' failed
make[2]: *** [all] Error 2
make[2]: Leaving directory '/home/lum3hc/work/github/qnx/dlt-daemon/qnx/build/nto-aarch64-le/build'
../common.mk:85: recipe for target 'dlt-daemon_all' failed
make[1]: *** [dlt-daemon_all] Error 2
make[1]: Leaving directory '/home/lum3hc/work/github/qnx/dlt-daemon/qnx/build/nto-aarch64-le'
recurse.mk:96: recipe for target 'all' failed
make: *** [all] Error 2
make: Leaving directory '/home/lum3hc/work/github/qnx/dlt-daemon/qnx/build'
```
Next steps: Finding rootcause, on analyzing

**Comment by minminlittleshrimp** [comment]
RCA: getline() introduced in QNX SDP 7.1. Current SDP 7.0
-> SDP to 7.1 + retry to make DLT QNX
https://www.qnx.com/developers/docs/7.1/#com.qnx.doc.neutrino.lib_ref/topic/g/getdelim.html

**Comment by michael-methner** [comment]
Hello @marcin-sochacki ,
patch looks good to me. 

we are facing build issue with sdp 7.0 as strndup() was introduced with 7.1 [1]. I think it could be merged anyways as this patch does not introduce the use of strndup().

Is there still any any need for sdp 7.0 support in the community?

@minminlittleshrimp : Please verfiy once on sdp 7.1 and merge if it works.

Sorry for taking long time for review. I will merge once we are able to compile and test your changes.

[1] https://www.qnx.com/developers/articles/rel_6778_0.html

**Comment by minminlittleshrimp** [comment]
Hello all,
I checked and it seems the submodule is deactivated somehow, and the make command always try to find gtest lib in /usr/local.
Hello @marcin-sochacki , can you check this?
```
-- Using pthread_setname_np API to set thread name
CMake Error at /usr/local/lib/cmake/GTest/GTestTargets.cmake:115 (message):
  The imported target "GTest::gtest" references the file

     "/usr/local/lib/libgtest.so.1.11.0"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/usr/local/lib/cmake/GTest/GTestTargets.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /usr/local/lib/cmake/GTest/GTestConfig.cmake:32 (include)
  /usr/share/cmake-3.16/Modules/FindGTest.cmake:178 (find_package)
  CMakeLists.txt:383 (find_package)
```
Plus the build of this patch is fine if I turn all unitest flag off:
```
-             -DWITH_DLT_INSTALLED_TESTS=ON \
-             -DWITH_DLT_UNIT_TESTS=ON \
+             -DWITH_DLT_INSTALLED_TESTS=OFF \
+             -DWITH_DLT_UNIT_TESTS=OFF \
```
Hence I have 2 options:
1. We can merge this patch but set the flags OFF
2. You can rework for the testing phase and we will review again for also that part and consider merging later

What is your choice?
Regards

File: qnx/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+There are several CMAKE variables set for the QNX build as recommended by the dlt-daemon project maintainers:
+* WITH_DLT_QNX_SYSTEM=ON
+* WITH_DLT_CXX11_EXT=ON
+* DLT_IPC=UNIX_SOCKET
+* WITH_DLT_ADAPTOR=ON
+* WITH_DLT_USE_IPv6=OFF
+* WITH_LIB_SHORT_VERSION=ON
+
+If you want to customize dlt-dameon features via CMAKE variables, you have to edit *qnx\build\common.mk* file and modify **CMAKE_ARGS** variable.
```
Comment (minminlittleshrimp) [inline qnx/README.md:20]: If you want to customize **dlt-dameon** features -> **dlt-daemon**

**Comment by minminlittleshrimp** [comment]
All installation into sdp rfs done successfully and checked, both aarch64 and x86_64:
```
Dr.Mint@:~/qnx_tools/qnx-sdp.7.1/sdp/target/qnx7/aarch64le
$ find . -name dlt*
./usr/local/include/dlt
./usr/local/include/dlt/dlt_common.h
./usr/local/include/dlt/dlt_user.h
./usr/local/include/dlt/dlt_user_macros.h
./usr/local/include/dlt/dlt_types.h
./usr/local/include/dlt/dlt_client.h
./usr/local/include/dlt/dlt_version.h
./usr/local/include/dlt/dlt_common_api.h
./usr/local/include/dlt/dlt_cpp_extension.hpp
./usr/local/include/dlt/dlt_offline_trace.h
./usr/local/include/dlt/dlt_multiple_files.h
./usr/local/include/dlt/dlt_filetransfer.h
./usr/local/include/dlt/dlt_protocol.h
./usr/local/include/dlt/dlt.h
./usr/local/include/dlt/dlt_shm.h
./usr/local/share/dlt-filetransfer
./usr/local/share/dlt-filetransfer/dlt-test-filetransfer-file
./usr/local/share/dlt-filetransfer/dlt-test-filetransfer-image.png
./usr/local/bin/dlt-test-fork-handler
./usr/local/bin/dlt-test-user
./usr/local/bin/dlt-convert
./usr/local/bin/dlt-sortbytimestamp
./usr/local/bin/dlt-test-stress-user
./usr/local/bin/dlt-example-filetransfer
./usr/local/bin/dlt-test-multi-process
./usr/local/bin/dlt-test-init-free
./usr/local/bin/dlt-daemon
./usr/local/bin/dlt-passive-node-ctrl
./usr/local/bin/dlt-test-filetransfer
./usr/local/bin/dlt-control
./usr/local/bin/dlt-test-client
./usr/local/bin/dlt-test-non-verbose
./usr/local/bin/dlt-example-user-func
./usr/local/bin/dlt-test-preregister-context
./usr/local/bin/dlt-logstorage-ctrl
./usr/local/bin/dlt-adaptor-udp
./usr/local/bin/dlt-test-stress-client
./usr/local/bin/dlt-test-cpp-extension
./usr/local/bin/dlt-test-stress
./usr/local/bin/dlt-adaptor-stdin
./usr/local/bin/dlt-test-multi-process-client
./usr/local/bin/dlt-example-user
./usr/local/bin/dlt-qnx-system
./usr/local/bin/dlt-example-user-common-api
./usr/local/bin/dlt-test-qnx-slogger
./usr/local/bin/dlt-receive
./usr/local/etc/dlt.conf
./usr/local/etc/dlt-slog2ctxt.json
./usr/local/etc/dlt-qnx-system.conf
./usr/local/etc/dlt_gateway.conf
```
```
Dr.Mint@:~/qnx_tools/qnx-sdp.7.1/sdp/target/qnx7/x86_64
$ find . -name dlt*
./usr/local/include/dlt
./usr/local/include/dlt/dlt_common.h
./usr/local/include/dlt/dlt_user.h
./usr/local/include/dlt/dlt_user_macros.h
./usr/local/include/dlt/dlt_types.h
./usr/local/include/dlt/dlt_client.h
./usr/local/include/dlt/dlt_version.h
./usr/local/include/dlt/dlt_common_api.h
./usr/local/include/dlt/dlt_cpp_extension.hpp
./usr/local/include/dlt/dlt_offline_trace.h
./usr/local/include/dlt/dlt_multiple_files.h
./usr/local/include/dlt/dlt_filetransfer.h
./usr/local/include/dlt/dlt_protocol.h
./usr/local/include/dlt/dlt.h
./usr/local/include/dlt/dlt_shm.h
./usr/local/share/dlt-filetransfer
./usr/local/share/dlt-filetransfer/dlt-test-filetransfer-file
./usr/local/share/dlt-filetransfer/dlt-test-filetransfer-image.png
./usr/local/bin/dlt-test-fork-handler
./usr/local/bin/dlt-test-user
./usr/local/bin/dlt-convert
./usr/local/bin/dlt-sortbytimestamp
./usr/local/bin/dlt-test-stress-user
./usr/local/bin/dlt-example-filetransfer
./usr/local/bin/dlt-test-multi-process
./usr/local/bin/dlt-test-init-free
./usr/local/bin/dlt-daemon
./usr/local/bin/dlt-passive-node-ctrl
./usr/local/bin/dlt-test-filetransfer
./usr/local/bin/dlt-control
./usr/local/bin/dlt-test-client
./usr/local/bin/dlt-test-non-verbose
./usr/local/bin/dlt-example-user-func
./usr/local/bin/dlt-test-preregister-context
./usr/local/bin/dlt-logstorage-ctrl
./usr/local/bin/dlt-adaptor-udp
./usr/local/bin/dlt-test-stress-client
./usr/local/bin/dlt-test-cpp-extension
./usr/local/bin/dlt-test-stress
./usr/local/bin/dlt-adaptor-stdin
./usr/local/bin/dlt-test-multi-process-client
./usr/local/bin/dlt-example-user
./usr/local/bin/dlt-qnx-system
./usr/local/bin/dlt-example-user-common-api
./usr/local/bin/dlt-test-qnx-slogger
./usr/local/bin/dlt-receive
./usr/local/etc/dlt.conf
./usr/local/etc/dlt-slog2ctxt.json
./usr/local/etc/dlt-qnx-system.conf
./usr/local/etc/dlt_gateway.conf
```

For the inatruction of setting up in official release at: https://github.com/qnx/dlt-daemon/blob/v2.18.10_QNX/qnx/test/README.md, the setup for gtest dlt binaries not work, maybe build system trying to find a dynamic libgtest???
```
Dr.Mint@:~/work/github/qnx/test_qnx_official/dlt-daemon ((HEAD detached at v2.18.10_QNX)) 
$ CPULIST=x86_64 make -C qnx/build JLEVEL=4
...
/home/lum3hc/qnx_tools/qnx-sdp.7.1/sdp/host/linux/x86_64/usr/bin/x86_64-pc-nto-qnx7.1.0-ld: /usr/local/lib/libgtest.a(gtest-all.cc.o): in function `testing::internal::MatcherBase<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::DescribeTo(std::ostream*) const':
gtest-all.cc:(.text._ZNK7testing8internal11MatcherBaseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE10DescribeToEPSo[_ZNK7testing8internal11MatcherBaseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE10DescribeToEPSo]+0x74): undefined reference to `std::basic_ostream<char, std::char_traits<char> >& std::operator<< <std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*)'
/home/lum3hc/qnx_tools/qnx-sdp.7.1/sdp/host/linux/x86_64/usr/bin/x86_64-pc-nto-qnx7.1.0-ld: /usr/local/lib/libgtest.a(gtest-all.cc.o): in function `testing::internal::MatcherBase<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::DescribeNegationTo(std::ostream*) const':
gtest-all.cc:(.text._ZNK7testing8internal11MatcherBaseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE18DescribeNegationToEPSo[_ZNK7testing8internal11MatcherBaseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE18DescribeNegationToEPSo]+0x74): undefined reference to `std::basic_ostream<char, std::char_traits<char> >& std::operator<< <std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*)'
/home/lum3hc/qnx_tools/qnx-sdp.7.1/sdp/host/linux/x86_64/usr/bin/x86_64-pc-nto-qnx7.1.0-ld: /usr/local/lib/libgtest.a(gtest-all.cc.o): in function `testing::internal::MatcherBase<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>::DescribeNegationTo(std::ostream*) const':
gtest-all.cc:(.text._ZNK7testing8internal11MatcherBaseIRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE18DescribeNegationToEPSo[_ZNK7testing8internal11MatcherBaseIRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE18DescribeNegationToEPSo]+0x74): undefined reference to `std::basic_ostream<char, std::char_traits<char> >& std::operator<< <std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*)'
/home/lum3hc/qnx_tools/qnx-sdp.7.1/sdp/host/linux/x86_64/usr/bin/x86_64-pc-nto-qnx7.1.0-ld: warning: creating a DT_TEXTREL in a shared object
tests/CMakeFiles/gtest_dlt_user.dir/build.make:87: recipe for target 'tests/gtest_dlt_user' failed
make[4]: *** [tests/gtest_dlt_user] Error 1
make[4]: Leaving directory '/home/lum3hc/work/github/qnx/test_qnx_official/dlt-daemon/qnx/build/nto-x86_64-o/build'
CMakeFiles/Makefile2:1360: recipe for target 'tests/CMakeFiles/gtest_dlt_user.dir/all' failed
make[3]: *** [tests/CMakeFiles/gtest_dlt_user.dir/all] Error 2
make[3]: Leaving directory '/home/lum3hc/work/github/qnx/test_qnx_official/dlt-daemon/qnx/build/nto-x86_64-o/build'
Makefile:140: recipe for target 'all' failed
make[2]: *** [all] Error 2
make[2]: Leaving directory '/home/lum3hc/work/github/qnx/test_qnx_official/dlt-daemon/qnx/build/nto-x86_64-o/build'
../common.mk:84: recipe for target 'dlt-daemon_all' failed
make[1]: *** [dlt-daemon_all] Error 2
make[1]: Leaving directory '/home/lum3hc/work/github/qnx/test_qnx_official/dlt-daemon/qnx/build/nto-x86_64-o'
recurse.mk:96: recipe for target 'all' failed
make: *** [all] Error 2
make: Leaving directory '/home/lum3hc/work/github/qnx/test_qnx_official/dlt-daemon/qnx/build'
```
Regards

**Comment by marcin-sochacki** [comment]
Please go ahead with merging (I added one more commit with discussed changes).

By the way:
I'm not able to reproduce your gtest related build problems here.
Perhaps the files generated by cmake ony your host are not consistent due to earlier builds with QNX SDP 7.0 ?
Could you please try to clean your build environment:

make -C qnx/build clean

and then retry:

CPULIST=x86_64 make -C qnx/build JLEVEL=4

**Review by michael-methner — APPROVED** [review]

**Pull request merged**